### PR TITLE
PAYARA-3988 Dynamic Docker Node Instances

### PR DESCRIPTION
**Pull request by Pandrex247**
Brings the Autonaming functionality introduced recently to Docker Nodes.

The behaviour around the creation and lifecycle of _Docker Instances_ has changed to be as follows:
* If the Docker image is not present on the Docker Node machine, Payara Server will now instruct the Docker REST API to pull the image.
* `--autoname` can be used with the `create-instance` command as described in the original [Dynamic Instances PR](https://github.com/payara/Payara/pull/4090) to deal with any name conflicts.
  *  If there is a conflict in the Docker **container** names, the instance will still fail to be created. This is unlikely to happen during normal operation, and should only really occur if the DAS and the Docker Node somehow get out of sync.
* The `stop-instance` command will now stop the Docker container
* The Docker containers are now started, stopped, and deleted based on their ID, rather than their name.
  * This allows Payara Server instances created within Docker Containers spawned manually or via scaling to register themselves with the DAS and retain the ability to be managed by the DAS.
* If a new container is created and a name conflict is detected with another Payara Server instance, a new instance will be created.
  * This should only happen if you create a container manually, or it is somehow duplicated via scaling.
* If a new container is created and no instance name is provided to it, a new instance with an auto-generated name will be created.
  * Again, this should only happen in cases where you've created a container manually or it has otherwise been scaled up.

**Please Note**: If creating containers manually or otherwise scaling them, the instances are still tied to a Docker Node, and so the containers will fail to start if the host name does not match what has been configured on the DAS (or if the node doesn't exist).

# Examples
## Asadmin Examples
This example shows that no new behaviour should be expected on-top of what is already described in https://github.com/payara/Payara/pull/4090.
```
asadmin -a create-instance --node docky1 insty1     >>>     insty1 created in a container named insty1
asadmin -a create-instance --node docky1 insty1     >>>     insty1-BemusedBass created in a container named insty1-BemusedBass
asadmin -a create-instance --node docky1            >>>     Capricious-Carp created in a container named Capricious-Carp
```

## Manual/Scaling Examples
This example demonstrates what would happen if an instance called _insty1_ already existed on the DAS.
```
docker container create --network host --mount type=bind,source="/home/anon/passwordfile.txt",target="/opt/payara/passwords/passwordfile.txt",readonly -e PAYARA_DAS_HOST=payaraDas -e PAYARA_DAS_PORT=4848 -e PAYARA_NODE_NAME=docky1 -e PAYARA_INSTANCE_NAME=insty1 payara/server-node:latest     >>>     Instance named insty1-GarishGoldfish created in a container named elastic_ganguly
```
This example demonstrates creating new instances and containers on a Docker Node (you must create & configure the Docker Node on the DAS first), the last showing that the _PAYARA_INSTANCE_NAME_ environment variable is now optional:
```
docker container create --network host --mount type=bind,source="/home/anon/passwordfile.txt",target="/opt/payara/passwords/passwordfile.txt",readonly -e PAYARA_DAS_HOST=payaraDas -e PAYARA_DAS_PORT=4848 -e PAYARA_NODE_NAME=docky1 -e PAYARA_INSTANCE_NAME=insty2 payara/server-node:latest     >>>     Instance named insty2 created in a container named elastic_ganguly

docker container create --network host --mount type=bind,source="/home/anon/passwordfile.txt",target="/opt/payara/passwords/passwordfile.txt",readonly -e PAYARA_DAS_HOST=payaraDas -e PAYARA_DAS_PORT=4848 -e PAYARA_NODE_NAME=docky1 -e PAYARA_INSTANCE_NAME=insty2 payara/server-node:latest     >>>     Instance named insty2-BamboozledBarrcuda created in a container named wonderful_yonath

docker container create --network host --mount type=bind,source="/home/anon/passwordfile.txt",target="/opt/payara/passwords/passwordfile.txt",readonly -e PAYARA_DAS_HOST=payaraDas -e PAYARA_DAS_PORT=4848 -e PAYARA_NODE_NAME=docky1 payara/server-node:latest     >>>     Instance named Magnanimous-Monkfish  created in a container named sleepy_elgamel
```

### CODE REVIEW

**Comment by Pandrex247** [comment]
Jenkins test please

File: appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java [hunk 1]
```diff
@@ -127,36 +208,36 @@ public void execute(AdminCommandContext adminCommandContext) {
...
+            if (responseStatus.getFamily().equals(Response.Status.Family.SUCCESSFUL)) {
+                // Read the container ID from the response
+                JsonObject jsonResponse = response.readEntity(JsonObject.class);
+
+                if (jsonResponse != null) {
+                    String dockerContainerId = jsonResponse.getString("Id");
+
+                    // Set the Docker container ID either to the ID of the container, or to the instance name if the the
+                    // ID can't be obtained
+                    if (dockerContainerId != null && !dockerContainerId.equals("")) {
```
Comment (rdebusscher) [inline appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: Maybe use String.isEmpty()? like

 !dockerContainerId.trim().isEmpty())

File: appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java [hunk 2]
```diff
@@ -103,6 +148,42 @@ public void execute(AdminCommandContext adminCommandContext) {
             return;
         }
 
+        // Pull the image if it hasn't been downloaded or built yet
+        pullImage(adminCommandContext, actionReport, node);
+
+        createContainer(adminCommandContext, actionReport, node, server, dasHost, dasPort);
+    }
+
+    private void pullImage(AdminCommandContext adminCommandContext, ActionReport actionReport, Node node) {
```
[the remark this answers is not in the document]
Comment (rdebusscher) [reply appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: Is it correct that pull Image uses the same Docker URL endpoint as createContainer.

Should it not be the _images/create_ URL?

**Comment by Pandrex247** [comment]
Jenkins test please

File: appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java [hunk 3]
```diff
@@ -127,36 +208,36 @@ public void execute(AdminCommandContext adminCommandContext) {
...
         // Check status of response and act on result
         if (response != null) {
             Response.StatusType responseStatus = response.getStatusInfo();
-            if (!responseStatus.getFamily().equals(Response.Status.Family.SUCCESSFUL)) {
+            if (responseStatus.getFamily().equals(Response.Status.Family.SUCCESSFUL)) {
+                // Read the container ID from the response
+                JsonObject jsonResponse = response.readEntity(JsonObject.class);
+
+                if (jsonResponse != null) {
+                    String dockerContainerId = jsonResponse.getString("Id");
```
Comment (jGauravGupta) [inline appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: `Id` key defined in lower case on StopDockerContainerCommand @RestParam(name="id") annotation.

File: appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java [hunk 4]
```diff
@@ -173,6 +254,55 @@ public void execute(AdminCommandContext adminCommandContext) {
         }
     }
 
+    private WebTarget createWebTarget(Node node, String endpoint) {
+        Client client = ClientBuilder.newClient();
+        WebTarget webTarget = null;
+        if (Boolean.parseBoolean(node.getUseTls())) {
```
Comment (AlanRoth) [inline appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: Correct me if I'm wrong but could you do something like this to reduce redundency?;
```
boolean isHttps = Boolean.parseBoolean(node.getUseTls())
if(endpoint.startsWith("/")) {
    webTarget = client.target(isHttps ? "https://" : "http://" 
                  + node.getNodeHost()
                  + ":"
                  + node.getDockerPort()
                  + endpoint);
} else {
    webTarget = client.target(isHttps ? "https://" : "http://"
                  + node.getNodeHost()
                  + ":"
                  + node.getDocketPort()
                  + "/"
                  + endpoint);
}
```

File: appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java (hunk 3 above)
Comment (Pandrex247) [reply appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: This is referencing the "Id" field from the Docker Rest API, not our own.
The Docker Rest API has the capital :)
https://docs.docker.com/engine/api/v1.39/#operation/ContainerCreate

File: appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java (hunk 4 above)
Comment (Pandrex247) [reply appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: I'll give it a test - if I'm going with ternary form I might as well go whole hog with something like the following:
```
WebTarget webTarget = client.target((Boolean.parseBoolean(node.getUseTls()) ? "https://" : "http://")
                + node.getNodeHost()
                + ":"
                + node.getDockerPort()
                + (endpoint.startsWith("/") ? endpoint : "/" + endpoint));
```

Comment (AlanRoth) [reply appserver/docker/src/main/java/fish/payara/docker/instance/admin/CreateDockerContainerCommand.java]: That works too, and still somewhat readable :P

**Comment by Pandrex247** [comment]
Jenkins test please

**Pull request merged**